File: ising_simulation.py
#Translated julia script to python
#In julia you can use greek symbols as variables and some other operators, you'll probably have to parse all these symbols to strings
import numpy as np
from numba import njit #Maybe you don't need this package, i was using multithreading to simulate one sequence on each core, there is probably a better way to do this in python

#There is probably a better way to do an ising simulation with Glauber dynamics in python

# Simulates the Ising model in 2D using Glauber dynamics.

# Generates a random 2D Ising configuration with dimensions m x n.
# Elements are randomly assigned -1 or 1.
def rand_ising2d(m, n=None):
    if n is None:
        n = m
    return np.random.choice([-1, 1], size=(m, n), dtype=np.int8)

# Set random seed for reproducibility.
np.random.seed(4649)

# Critical inverse temperature for the 2D Ising model.
# β (beta) is replaced with beta_crit
beta_crit = np.log(1 + np.sqrt(2)) / 2

# Precompute probabilities for spin flips at the critical temperature.
# This avoids repeated calculations within the evolution function.
prob = np.exp(-2 * beta_crit * np.arange(-4, 5))

# Evolves a 2D Ising configuration `s` for `niters` iterations at inverse temperature `beta`.
# β (beta) is replaced with beta
@njit  # Use Numba for JIT compilation and speedup
def ising2d_evolve(s, beta, niters):
    m, n = s.shape  # Get dimensions of the lattice
    prob = np.exp(-2 * beta * np.arange(-4, 5))  # Probabilities for spin flips
    for _ in range(niters):  # Loop over iterations
        for j in range(n):
            for i in range(m):
                # Get the neighboring spins with periodic boundary conditions
                NN = s[(i - 1) % m, j]
                SS = s[(i + 1) % m, j]
                WW = s[i, (j - 1) % n]
                EE = s[i, (j + 1) % n]
                CT = s[i, j]
                # Calculate the energy change if the spin is flipped
                k = CT * (NN + SS + WW + EE)
                # Flip the spin with probability prob[k+4] (adjusting index for Python)
                if np.random.rand() < prob[k + 4]:
                    s[i, j] *= -1
    return s

# Generates an initial ensemble of M random Ising configurations with dimensions N x N.
def isingInitialEnsemble(M, N):
    return [rand_ising2d(N) for _ in range(M)]

# Generates a dataset of M Ising configurations, each evolving for n_frames frames.
# Starts with an initial random configuration and quenches to inverse temperature `beta`.
# β (beta) is replaced with beta
def isingDataSet(M, N, beta, n_frames, n_t=1):
    X = []  # Initialize an empty list to store configurations
    X0 = isingInitialEnsemble(M, N)  # Generate initial random configurations

    for i in range(M):
        X.append(X0[i].copy())  # Add the initial configuration to the dataset
        for _ in range(1, n_frames):
            X.append(X[-1].copy())  # Add a copy of the previous configuration
            X[-1] = ising2d_evolve(X[-1], beta, n_t)  # Evolve the configuration for n_t steps

    return X

# Computes the distance matrix between Ising configurations in A.
# The distance between two configurations is the sum of squared differences between their spins.
def isingDistanceMatrix(A):
    M = len(A)
    D = np.zeros((M, M))
    for i in range(M):
        for j in range(i):  # Python uses 0-based indexing
            D[i, j] = np.sum((A[i] - A[j]) ** 2)  # Calculate the distance
            D[j, i] = D[i, j]  # Ensure symmetry
    return D

# Computes the Diffusion Map distance matrix between Ising configurations in A.
# Uses the exponential of the negative squared distance scaled by epsilon.
# ϵ (epsilon) is replaced with epsilon
def isingDiffMapDistance(A, M, N, epsilon):
    D = np.zeros((M, M))  # Initialize the distance matrix
    for i in range(M):
        for j in range(M):
            # Calculate the Diffusion Map distance
            D[i, j] = np.exp(-isingDistance(A[i], A[j]) / ((2 * N) ** 2 * epsilon))  
    return D

# Generates a dataset of M Ising configurations, each quenched to inverse temperature `beta`.
# Only returns the final state of each realization.
# β (beta) is replaced with beta
def isingSteadyDataSet(M, N, beta, n_t):
    X = isingInitialEnsemble(M, N)  # Generate initial random configurations
    for x in X:
        x = ising2d_evolve(x, beta, n_t)  # Evolve the configuration for n_t steps
    return X

File: ising_simulation.jl
rand_ising2d(m, n=m) = rand(Int8[-1, 1], m, n)

Random.seed!(4649)

const β_crit = log(1+sqrt(2))/2
prob = @SVector [exp(-2*β_crit*k) for k in -4:4]

function ising2d_evolve!(s, β, niters; rng=default_rng())
    m, n = size(s)
    prob = @SVector [exp(-2*β*k) for k in -4:4]
    @fastmath @inbounds @simd for iter in 1:niters
        for j in 1:n 
            for i in 1:m
                let NN = s[ifelse(i == 1, m, i-1), j],
                    SS = s[ifelse(i == m, 1, i+1), j],
                    WW = s[i, ifelse(j == 1, n, j-1)],
                    EE = s[i, ifelse(j == n, 1, j+1)],
                    CT = s[i, j]
                    k = CT * (NN + SS + WW + EE)
                    s[i,j] = ifelse(rand(rng) < prob[k+5], -CT, CT)
                end
            end
        end
    end
end

function isingE(A,N;J=1)
    E=0
    for i in 0:N-1
        for j in 0:N-1
            E += -J*A[i+1,j+1]*(A[mod(i+1,N)+1,j+1]+A[mod(i-1,N)+1,j+1]+A[i+1,mod(j+1,N)+1]+A[i+1,mod(j-1,N)+1])#Nearest neighbours energy
        end
    end
    return E/2
end

function isingInitialEnsemble(M,N)
    return [rand_ising2d(N) for i in 1:M]
end

function isingDataSet(M,N,β,n_frames;n_t=1,rng=Random.default_rng())

    X=Vector{Matrix{Int64}}(undef,0)
    X0=isingInitialEnsemble(M,N)

    for i in 1:M
        push!(X,deepcopy(X0[i]))
        for j in 2:n_frames
            push!(X,deepcopy(X[end]))
            ising2d_evolve!(X[end],β,n_t;rng)
        end
    end

    return X
end

function isingDistanceMatrix(A)
    M = length(A)
    D=zeros(M,M)
    @sync for i in 1:M
        @spawn for j in 1:i-1
            D[i,j]=sum((A[i]-A[j]).^2)
            D[j,i]=D[i,j]
        end
    end
    return D
end

function isingDiffMapDistance(A,M,N,ϵ)
    
    # p=Progress(M*M)
    @sync for i in 1:M
        @spawn for j in 1:M
            D[i,j]=exp(-isingDistance(A[i],A[j])/((2*N)^2*ϵ))
            # next!(p)
        end
    end
end

function isingSteadyDataSet(M,N,β,n_frames,n_t;rng=default_rng())
    # X=[ising0condition(N)*(-1)^i for i in 1:M]
    X=isingInitialEnsemble(M,N)
    for x in X
        ising2d_evolve!(x,β,n_t;rng)
    end
    
    return X
end

function IsingSGEMinima(M,N,n_frames,Ts; n_t = 1, step = 10) #Fixed N data set for variable 
    minima = Vector{Vector{Int}}(undef,0)
    for T in Ts
        β = 1. / T
        X = isingDataSet(M,N,β,n_frames, n_t = n_t) #Data set at temperature T
        D = isingDistanceMatrix(X)
        ϵs = (0.00001:0.00001:0.01)∪(0.0101:0.0001:0.05)∪(0.051:0.001:0.1)∪(0.11:0.01:1)
        SGE = SGECriteria(D[1:step:end,1:step:end],ϵs)
        indices, minima_T = findminima(SGE)
    end
end

